File: backend/app/domain/filters.py
"""The shared filter model.

One of the load-bearing product decisions: the employee list and the analytics
dashboard consume the *same* filter object. A number on the dashboard is
therefore always the aggregate of a list the HR Manager can navigate to and
inspect — "€4.2M in Germany" is one click away from the 340 people it is the sum
of. Two separate filter models would let those two views drift apart, and an
unverifiable dashboard is worse than no dashboard.

The model is a frozen dataclass rather than a Pydantic model because it is a
domain value, not a wire format: the HTTP layer parses query parameters into it,
and the repositories translate it into SQL predicates. Neither layer owns it.
"""

from dataclasses import dataclass
from typing import Final, Self

from app.domain.enums import Department, EmploymentStatus, EmploymentType, Level

DEFAULT_STATUSES: Final[tuple[EmploymentStatus, ...]] = (EmploymentStatus.ACTIVE,)
"""The selection an unqualified question implies.

"What do we pay engineers?" means the people on the payroll now. Former
employees stay one filter away, but including them by default would silently
inflate every headline figure.
"""


@dataclass(frozen=True, slots=True)
class EmployeeFilters:
    """A selection over the employee population.

    Every collection field is conjunctive across fields and disjunctive within
    one: ``countries=("DE", "FR")`` with ``departments=("SALES",)`` selects sales
    employees in Germany *or* France. An empty collection means "no constraint on
    this field", which is the natural reading of an untouched filter control.
    """

    search: str | None
    countries: tuple[str, ...]
    departments: tuple[Department, ...]
    levels: tuple[Level, ...]
    employment_types: tuple[EmploymentType, ...]
    statuses: tuple[EmploymentStatus, ...]
    min_total_comp_usd_minor: int | None
    max_total_comp_usd_minor: int | None

    def __post_init__(self) -> None:
        lower = self.min_total_comp_usd_minor
        upper = self.max_total_comp_usd_minor
        if lower is not None and lower < 0:
            raise ValueError(f"minimum compensation must be non-negative, got {lower}")
        if upper is not None and upper < 0:
            raise ValueError(f"maximum compensation must be non-negative, got {upper}")
        if lower is not None and upper is not None and lower > upper:
            raise ValueError(
                f"minimum compensation {lower} exceeds maximum {upper}; "
                "the range would select no employees"
            )

    @classmethod
    def unfiltered(cls) -> Self:
        """The whole population — the dashboard's default view."""
        return cls(
            search=None,
            countries=(),
            departments=(),
            levels=(),
            employment_types=(),
            statuses=(),
            min_total_comp_usd_minor=None,
            max_total_comp_usd_minor=None,
        )

    @property
    def is_empty(self) -> bool:
        """Whether this selects the whole population, used to label the UI."""
        return self == type(self).unfiltered()

File: frontend/src/lib/filters.ts
/**
 * The client half of the shared filter model.
 *
 * Filter state lives in the URL rather than in React state, for three reasons:
 *
 * 1.  A filtered view becomes a link. "Here's what we pay engineers in Germany"
 *     is something an HR Manager sends to a colleague, and that is only
 *     possible if the selection is in the address bar.
 * 2.  Moving between the dashboard and the employee list carries the selection
 *     across for free, which is what makes "these 340 people" clickable.
 * 3.  The parameter names are exactly the ones the API accepts, so the query
 *     string is passed to the backend verbatim. There is no second mapping that
 *     could disagree with the first.
 *
 * Compensation bounds are the one exception: the UI works in whole dollars
 * because that is how people think about salary, while the API takes minor
 * units. That conversion is explicit here and nowhere else.
 */

import { majorToMinor, minorToMajor } from './money'

export interface FilterState {
  search: string
  countries: string[]
  departments: string[]
  levels: string[]
  employmentTypes: string[]
  statuses: string[]
  /** Whole USD, not minor units — this is user-facing input. */
  minTotalCompUsd: number | null
  maxTotalCompUsd: number | null
}

export const EMPTY_FILTERS: FilterState = {
  search: '',
  countries: [],
  departments: [],
  levels: [],
  employmentTypes: [],
  statuses: [],
  minTotalCompUsd: null,
  maxTotalCompUsd: null,
}

/**
 * The default view: current employees only.
 *
 * An HR Manager asking "what do we pay?" means the people on the payroll now.
 * Former employees remain in the data and are one click away, but including
 * them by default would silently inflate every headline figure.
 */
export const DEFAULT_FILTERS: FilterState = { ...EMPTY_FILTERS, statuses: ['ACTIVE'] }

const PARAM = {
  search: 'search',
  country: 'country',
  department: 'department',
  level: 'level',
  employmentType: 'employment_type',
  status: 'status',
  minComp: 'min_total_comp_usd_minor',
  maxComp: 'max_total_comp_usd_minor',
} as const

/** Parameters that belong to the view, not the selection. */
const VIEW_PARAMS = new Set(['page', 'page_size', 'sort', 'direction', 'dimension'])

function parseOptionalDollars(raw: string | null): number | null {
  if (raw === null || raw.trim() === '') {
    return null
  }
  const minorUnits = Number(raw)
  return Number.isFinite(minorUnits) ? minorToMajor(minorUnits) : null
}

export function filtersFromSearchParams(params: URLSearchParams): FilterState {
  return {
    search: params.get(PARAM.search) ?? '',
    countries: params.getAll(PARAM.country),
    departments: params.getAll(PARAM.department),
    levels: params.getAll(PARAM.level),
    employmentTypes: params.getAll(PARAM.employmentType),
    statuses: params.getAll(PARAM.status),
    minTotalCompUsd: parseOptionalDollars(params.get(PARAM.minComp)),
    maxTotalCompUsd: parseOptionalDollars(params.get(PARAM.maxComp)),
  }
}

/**
 * Serialises a selection to the query string the API accepts.
 *
 * Empty values are omitted rather than sent as blanks, which keeps shared links
 * readable and means an untouched filter contributes nothing to the request.
 */
export function filtersToSearchParams(filters: FilterState): URLSearchParams {
  const params = new URLSearchParams()

  if (filters.search.trim() !== '') {
    params.set(PARAM.search, filters.search.trim())
  }
  for (const country of filters.countries) params.append(PARAM.country, country)
  for (const department of filters.departments) params.append(PARAM.department, department)
  for (const level of filters.levels) params.append(PARAM.level, level)
  for (const type of filters.employmentTypes) params.append(PARAM.employmentType, type)
  for (const status of filters.statuses) params.append(PARAM.status, status)

  if (filters.minTotalCompUsd !== null) {
    params.set(PARAM.minComp, String(majorToMinor(filters.minTotalCompUsd)))
  }
  if (filters.maxTotalCompUsd !== null) {
    params.set(PARAM.maxComp, String(majorToMinor(filters.maxTotalCompUsd)))
  }

  return params
}

/** Merges a selection into existing params, leaving view params untouched. */
export function mergeFiltersIntoParams(
  filters: FilterState,
  existing: URLSearchParams,
): URLSearchParams {
  const merged = filtersToSearchParams(filters)
  for (const [key, value] of existing.entries()) {
    if (VIEW_PARAMS.has(key)) {
      merged.append(key, value)
    }
  }
  return merged
}

/** How many facets are narrowing the view, for the "clear filters" affordance. */
export function activeFilterCount(filters: FilterState): number {
  return [
    filters.search.trim() !== '',
    filters.countries.length > 0,
    filters.departments.length > 0,
    filters.levels.length > 0,
    filters.employmentTypes.length > 0,
    filters.statuses.length > 0,
    filters.minTotalCompUsd !== null,
    filters.maxTotalCompUsd !== null,
  ].filter(Boolean).length
}

/**
 * Whether the range is self-contradictory.
 *
 * The API rejects an inverted range with a 400, but catching it here lets the
 * form say so next to the field instead of surfacing a failed request.
 */
export function compensationRangeError(filters: FilterState): string | null {
  const { minTotalCompUsd, maxTotalCompUsd } = filters
  if (minTotalCompUsd !== null && maxTotalCompUsd !== null && minTotalCompUsd > maxTotalCompUsd) {
    return 'Minimum is above the maximum, so no one would match'
  }
  return null
}
